Fixed Button und ErsterVormerker

Benokit4 opened the pull request:
Button ist nur anwählbar, wenn Ausleiher der erste Vormerker ist oder es keinen Vormerker gibt.

Erster Vormerker wird nach Ausleihe entfernt, nachfolgende rutschen nach.

Benokit4 committed:
Fixed Button und ErsterVormerker

Button ist nur anwählbar, wenn Ausleiher der erste Vormerker ist oder es keinen Vormerker gibt.

Erster Vormerker wird nach Ausleihe entfernt, nachfolgende rutschen nach.

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/vormerk/VormerkServiceImpl.java:
```diff
@@ -33,6 +38,11 @@ public void merkeVor(Kunde kunde, Medium medium)
+        {
+        	vormerkkarte = new Vormerkkarte(medium);
+        	_vormerkkarten.put(medium, vormerkkarte);
+        }
```
Kann man weglassen, da getVormerkkarte nie null zurückgibt

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/vormerk/VormerkServiceImpl.java:
```diff
@@ -49,13 +59,19 @@ public boolean istVormerkenMoeglich(Kunde kunde, Medium medium)
         Vormerkkarte vormerkkarte = getVormerkkarte(medium);
         
-        return vormerkkarte.istVormerkenMoeglich(kunde);
+        return vormerkkarte == null || vormerkkarte.istVormerkenMoeglich(kunde);
```
vormerkkarte kann nie null sein, also false || x = x

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/vormerk/VormerkServiceImpl.java:
```diff
@@ -49,13 +59,19 @@ public boolean istVormerkenMoeglich(Kunde kunde, Medium medium)
     public Vormerkkarte getVormerkkarte(Medium medium)
     {
-        return _vormerkkarten.computeIfAbsent(medium, (x) -> new Vormerkkarte(x));
+    	Vormerkkarte vormerkkarte = _vormerkkarten.get(medium);
```
Macht das gleiche, ist aber (noch) verständlicher

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/vormerk/VormerkServiceImpl.java:
```diff
@@ -88,7 +104,7 @@ private boolean entferneErstenVormerker(Kunde kunde, Medium medium)
         
         Vormerkkarte vormerkkarte = getVormerkkarte(medium);
-        if (vormerkkarte.getErsterVormerker() == kunde)
+        if (vormerkkarte != null && kunde.equals(vormerkkarte.getErsterVormerker()))
```
vormerkkarte != null ist Redundant (immer true), da true && x = x
kunde.equals(vormerkkarte.getErsterVormerker()) ist besser als zuvor, ermöglicht mehr Abstraktion

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/vormerk/VormerkServiceImpl.java:
```diff
@@ -98,4 +114,14 @@ private boolean entferneErstenVormerker(Kunde kunde, Medium medium)
+    public void entferneVormerker(Medium medium, Kunde kunde)
+    {
+    	Vormerkkarte vormerkkarte = getVormerkkarte(medium);
+    	if (vormerkkarte != null && kunde.equals(vormerkkarte.getErsterVormerker()))
```
vormerkkarte != null ist redundant, da immer true, und true && x = x

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/startup/StartUpMediathek_Blatt06.java:
```diff
@@ -70,9 +70,10 @@ private static void erstelleServices()
-                    _medienbestand, datenEinleser.getVerleihkarten());
-            _vormerkService = new VormerkServiceImpl(_verleihService);
+                    _medienbestand, datenEinleser.getVerleihkarten(), _vormerkService);
+            ((VormerkServiceImpl)_vormerkService).setVerleihService(_verleihService);;
```
Cast auf untertyp ist sehr blöd, wenn dynamischer Typ anders, stürzt das Programm ab

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/ui/subwerkzeuge/ausleihemedienauflister/AusleiheMedienauflisterWerkzeug.java:
```diff
@@ -91,8 +92,14 @@ private void setzeAnzuzeigendeMedien()
+            Kunde ersterVormerker = null;
+            Vormerkkarte vormerkkarte = _vormerkService.getVormerkkarte(medium);
 
+            if (vormerkkarte != null)
```
vormerkkarte != null ist immer true, daher kann man die Bedingung weglassen

frndotmsn replied:
Reihenfolge ändern, ersterVormerker = vormerkkarte.getErsterVormerker()

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/vormerk/VormerkService.java:
```diff
@@ -70,4 +70,11 @@ public interface VormerkService extends ObservableService
      */
     public Vormerkkarte getVormerkkarte(Medium medium);
+    
+    /**
```
Gibts schon in VormerkService#entferneErstenVormerker

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/vormerk/VormerkServiceImpl.java:
```diff
@@ -16,13 +16,18 @@
-    public VormerkServiceImpl(VerleihService verleihService)
+    public VormerkServiceImpl()
     {
-        _verleihService = verleihService;
```
Ändeungen rückgangig machen, man braucht keine Setter-Methode, geht auch über den Konstruktor und als final-Feld!

frndotmsn reviewed (changes requested):
Viele gute Ideen, aber mehr Nachbedingungen anschauen & Änderungen aus https://github.com/frndotmsn/SE2Laborphase2/pull/39 anschaun

frndotmsn commented:
kümmert sich um https://github.com/frndotmsn/SE2Laborphase2/issues/40 und um https://github.com/frndotmsn/SE2Laborphase2/issues/36

Benokit4 committed:
Verbesserungen nach Kommentaren vorgenommen

Leider funktioniert es nun nicht mehr, dass der Vormerker nach der Ausleihe entfernt wird. Bekomme es auch nicht gefixed.

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/vormerk/VormerkService.java:
```diff
@@ -51,13 +51,13 @@ public interface VormerkService extends ObservableService
      * Entfernt die ersten Vormerker der Medien falls dieser dem Kunden entspricht.
      * Falls es keinen ersten Vormerker gibt oder der Kunde diesem nicht entspricht passiert nichts.
      * 
+     * @param medium die Medien
```
Lieber den Namen medien so lassen, ist ja ne Liste von Medien und nicht nur ein Medium

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/vormerk/VormerkServiceImpl.java:
```diff
@@ -55,18 +55,28 @@ public boolean istVormerkenMoeglich(Kunde kunde, Medium medium)
     
     @Override
-    public void entferneErstenVormerker(Kunde kunde, List<Medium> medien)
+    public void entferneErstenVormerker(List<Medium> medien, Kunde kunde)
```
Parameterreihenfolge kann wie vorher bleiben

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/vormerk/VormerkServiceImpl.java:
```diff
@@ -55,18 +55,28 @@ public boolean istVormerkenMoeglich(Kunde kunde, Medium medium)
         
         for (Medium medium : medien)
         {
-            entferneErstenVormerker(kunde, medium);
```
Warum nicht einfach entferneErstenVormerker aufrufen?

frndotmsn commented on src/test/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/verleih/VerleihServiceImplTest.java:
```diff
@@ -55,8 +58,13 @@ public VerleihServiceImplTest()
         medium = new CD("CD4", "baz", "foo", 123);
         medienbestand.fuegeMediumEin(medium);
         _medienListe = medienbestand.getMedien();
+        
```
Das wird so nicht klappen, in Zeile 59 gibst du die momentane Referenz an VerleihServiceImpl weiter, jedoch setzt du in Zeile _vormerkService auf eine neue Instanz, _service zeigt jedoch noch auf die alte!
Das ist so wie
x = new FalschesX()
y = new Y(x)
x = new RichtigesX()

y zeigt immer noch auf das falsche X!

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/startup/StartUpMediathek_Blatt06.java:
```diff
@@ -70,9 +70,12 @@ private static void erstelleServices()
-            _verleihService = new VerleihServiceImpl(_kundenstamm,
-                    _medienbestand, datenEinleser.getVerleihkarten());
-            _vormerkService = new VormerkServiceImpl(_verleihService);
+            VormerkServiceImpl vormerkServiceImpl = new VormerkServiceImpl(null); // temporär null
```
Das wird so nicht klappen, in Zeile 59 gibst du die momentane Referenz an VerleihServiceImpl weiter, jedoch setzt du in Zeile _vormerkService auf eine neue Instanz, _service zeigt jedoch noch auf die alte!
Das ist so wie
x = new FalschesX()
y = new Y(x)
x = new RichtigesX()

y zeigt immer noch auf das falsche X!

frndotmsn replied:
Doch ggf. Tests für StartUpMediathek???
Methode für das Erstellen der Services?

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/ui/rueckgabe/VerleihkartenTableModel.java:
```diff
@@ -40,6 +40,18 @@ public VerleihkartenTableModel()
     {
         _verleihkartenListe = new ArrayList<Verleihkarte>();
     }
+    
```
Lieber bisherigen Konstruktor verwenden und danach setVerleihkarten aufrufen, dies Triggert nämlich fireOnDataChanged, man hätte das im Konstruktor machen können, sofern sich die Verleihkarten nie ändern!

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/ui/subwerkzeuge/ausleihemedienauflister/AusleiheMedienauflisterWerkzeug.java:
```diff
@@ -91,8 +92,11 @@ private void setzeAnzuzeigendeMedien()
             // der Name des Vormerkers, an den ein Medium ausgeliehen werden
             // darf, gemäß Anforderung c).
-            Kunde ersterVormerker = _vormerkService.getVormerkkarte(medium).getErsterVormerker();
+            Kunde ersterVormerker = null;
```
Also im Endeffekt ist ersterVormerker immer nur _vormerkkarte.getErsterVormerker() = _vormerkService.getVormerkkarte(medium).getErsterVormerker()!
Da wir nirgends sonst die vormerkkarte brauchen, können wir es so lassen wie zuvor!

frndotmsn commented on src/test/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/vormerk/VormerkServiceImplTest.java:
```diff
@@ -64,9 +64,8 @@ public VormerkServiceImplTest()
-        _verleihService = new VerleihServiceImpl(kundenstamm, medienbestand, new ArrayList<Verleihkarte>());
-        
-        
+        _vormerkService = new VormerkServiceImpl(null);//temporär null
```
Das wird so nicht klappen, in Zeile 59 gibst du die momentane Referenz an VerleihServiceImpl weiter, jedoch setzt du in Zeile _vormerkService auf eine neue Instanz, _service zeigt jedoch noch auf die alte!
Das ist so wie
x = new FalschesX()
y = new Y(x)
x = new RichtigesX()

y zeigt immer noch auf das falsche X!

frndotmsn commented on src/test/java/de/uni_hamburg/informatik/swt/se2/mediathek/ui/rueckgabe/VerleihkartenTableModelTest.java:
```diff
@@ -51,10 +53,12 @@ public VerleihkartenTableModelTest()
         verleihkarten.add(_karte1);
         verleihkarten.add(_karte2);
+        
+        VormerkService vormerkService = new VormerkServiceImpl(null); //temporär null
```
Wird so nicht klappen, zeigt noch auf die alte Instanz!

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/vormerk/VormerkServiceImpl.java:
```diff
@@ -98,4 +108,14 @@ private boolean entferneErstenVormerker(Kunde kunde, Medium medium)
             return false;
         }
     }
+    
```
Duplikat von entferneErstenVormerker!

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/vormerk/VormerkServiceImpl.java:
```diff
@@ -98,4 +108,14 @@ private boolean entferneErstenVormerker(Kunde kunde, Medium medium)
+    	if (kunde.equals(vormerkkarte.getErsterVormerker()))
+    	{
+    		vormerkkarte.entferneErstenVormerker();
+    		informiereUeberAenderung();
```
Wir wollen das nirgendswo anders außer in entferneErsteVormerker aufrufen, wieso dann public?

frndotmsn replied:
Unnötiges informiereUeberAenderung, falls private, starke Performanceeinbuße, falls wir das für jede veränderung machen, und nicht nur falls wir das nach allen änderungen machen!

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/vormerk/VormerkService.java:
```diff
@@ -51,13 +51,13 @@ public interface VormerkService extends ObservableService
      * @require medien != null
      */
-    public void entferneErstenVormerker(Kunde kunde, List<Medium> medien);
+    public void entferneErstenVormerker(List<Medium> medium, Kunde kunde);
```
Parameter so wie zuvor lassen, der neue name für medien macht weniger sinn

frndotmsn commented on src/main/java/de/uni_hamburg/informatik/swt/se2/mediathek/services/verleih/VerleihServiceImpl.java:
```diff
@@ -39,6 +40,11 @@ public class VerleihServiceImpl extends AbstractObservableService
      * Der Kundenstamm.
      */
     private KundenstammService _kundenstamm;
+    
```
Es wäre einfacher, wenn beide nicht Referenzen aufeinander bräuchten! (Das ließe sich dann durch zusammenlegung von VormerkService & VerleihService am besten regeln, würde aber zu zu komplexen Code führen!)
Daher: brauchen wir wirklich vormerkService hier?
Idee: vormerkService als Parameter an verleiheAn übergeben, wenn wir an den Punkten an denen wir verleiheAn aufrufen eine Referenz auf vormerkService haben!

frndotmsn reviewed (changes requested):
Kommentare wurden versucht umzusetzen, es stellt sich heraus dass der zirkuläre referenz zwischen VormerkService & VerleihService NICHT gut ist (ist dann auch noch schwieriger für den Garbage-Collector!)

frndotmsn commented:
Nächstes Mal am besten auf vor den Commit zurücksetzen, nochmal versuchen und dabei an dem Commit und den Kommentaren orientieren

frndotmsn commented:
Ich werds jetzt kurz versuchen

frndotmsn committed:
VormerkService & VerleihService entkoppelt, #36 und #40 implementiert

VormerkServiceImpl und VerleihServiceImpl halten nichtmehr Referezen aufeinander, stattdessen muss bein Aufruf von bestimmten Methoden von vormerkService und VerleihService der jeweils andere Service übergeben werden, dies hätte sich auch durch eine Setter-Methode nach der Instanziierung eines Services realisieren lassen, das Downcasting eines Services in einen Subtyp sollte vermieden werden. Der "ausleihen"-Knopf in der Ausleih-Ansicht lässt sich nun nur betätigen, sofern alle Medien auch nach den Kriterien des Vormerkens ausgeliehen werden dürfen. Wird ein Medium, welches vorgemerkt war, ausgeliehen, so wird der erste Vormerker, der nun der Ausleiher ist, entfernt. Die restlichen Vormerker des Mediums rücken nach.

frndotmsn committed:
spotless hinzugefügt mit SE2Format.xml

frndotmsn committed:
Merge branch 'aufgabe-6-ausleihen-knopf-und-entfernen-des-vormerkers' into Aufgabe-6-fixed-Knopf-und-ErsterVormerker

frndotmsn committed:
Formatting mit spotless zu GitHub-Actions hinzugefügt

frndotmsn reviewed (approved):
LGTM!
